Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -0,0 +1,252 @@
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "config.h"
+#include "WaiterListManager.h"
+
+#include "JSGlobalObject.h"
+#include "JSLock.h"
+#include "ObjectConstructor.h"
+#include <wtf/DataLog.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/RawPointer.h>
+
+namespace JSC {
+
+namespace WaiterListsManagerInternal {
+static constexpr bool verbose = false;
+}
+
+WaiterListManager& WaiterListManager::singleton()
+{
+    static LazyNeverDestroyed<WaiterListManager> manager;
+    static std::once_flag onceKey;
+    std::call_once(onceKey, [&] {
+        manager.construct();
+    });
+    return manager;
+}
+
+JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr, AtomicsWaitValidation validation, Seconds timeout, AtomicsWaitType type)
+{
+    if (type == AtomicsWaitType::Async) {
+        JSObject* object = constructEmptyObject(globalObject->vm(), globalObject->objectStructureForObjectConstructor());
+
+        bool isAsync = false;
+        JSValue value;
+        if (validation == AtomicsWaitValidation::Fail)
+            value = vm.smallStrings.notEqualString();
+        else if (!timeout)
+            value = vm.smallStrings.timedOutString();
+        else {
+            isAsync = true;
+            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+            addAsyncWaiter(ptr, promise, timeout);
+            value = promise;
+        }
+
+        object->putDirect(vm, vm.propertyNames->async, jsBoolean(isAsync));
+        object->putDirect(vm, vm.propertyNames->value, value);
+        return object;
+    }
+
+    if (validation == AtomicsWaitValidation::Fail)
+        return vm.smallStrings.notEqualString();
+
+    Waiter waiter { &vm };
+    WaiterList* list = findOrCreateList(ptr);
+    const TimeWithDynamicClockType& time = MonotonicTime::now() + timeout;
```

Comment:
Do not use const reference here, it is incorrect, right? Just having reference is wrong in C++.
You should use `MonotonicTime time = MonotonicTime::now() + timeout`.